src/main/scala/com.snowplowanalytics.snowplow.analytics.scalasdk/Event.scala

```diff
@@ -0,0 +1,218 @@
+                 os_timezone:              Option[String],
+                 dvce_type:                Option[String],
+                 dvce_ismobile:            Option[Boolean],
+                 dvce_screenwidth:         Option[Int],
+                 dvce_screenheight:        Option[Int],
+                 doc_charset:              Option[String],
+                 doc_width:                Option[Int],
+                 doc_height:               Option[Int],
+                 tr_currency:              Option[String],
+                 tr_total_base:            Option[Double],
+                 tr_tax_base:              Option[Double],
+                 tr_shipping_base:         Option[Double],
+                 ti_currency:              Option[String],
+                 ti_price_base:            Option[Double],
+                 base_currency:            Option[String],
+                 geo_timezone:             Option[String],
+                 mkt_clickid:              Option[String],
+                 mkt_network:              Option[String],
+                 etl_tags:                 Option[String],
+                 dvce_sent_tstamp:         Option[Instant],
+                 refr_domain_userid:       Option[String],
+                 refr_device_tstamp:       Option[Instant],
+                 derived_contexts:         Contexts,
+                 domain_sessionid:         Option[String],
+                 derived_tstamp:           Option[Instant],
+                 event_vendor:             Option[String],
+                 event_name:               Option[String],
+                 event_format:             Option[String],
+                 event_version:            Option[String],
+                 event_fingerprint:        Option[String],
+                 true_tstamp:              Option[Instant]) {
+
+  def inventory: Set[Data.ShreddedType] = {
+    val unstructEvent = unstruct_event
+      .data
+      .toSet
+      .map((ue: SelfDescribingData[Json]) => Data.ShreddedType(Data.UnstructEvent, ue.schema))
+
+    val derivedContexts = derived_contexts
+      .data
+      .toSet
+      .map((ctx: SelfDescribingData[Json]) => Data.ShreddedType(Data.Contexts(Data.DerivedContexts), ctx.schema))
+
+    val customContexts = contexts
+      .data
+      .toSet
+      .map((ctx: SelfDescribingData[Json]) => Data.ShreddedType(Data.Contexts(Data.CustomContexts), ctx.schema))
+
+    customContexts ++ derivedContexts ++ unstructEvent
+  }
+
+
+  def atomic(): List[(String, Json)] =
+    this.asJsonObject.toList.flatMap {
+      case ("contexts", _)         => None
+      case ("unstruct_event", _)   => None
+      case ("derived_contexts", _) => None
+      case (k, v) => Some((k, v))
+    }
+
+  def toJson(lossy: Boolean): Json =
+    if (lossy) {
+      // Convert initial JSON to map, filter out auto-derived contexts and transform unstruct_event to JSON
+      val transformed = this.asJsonObject.toMap
```

Comment:
Good idea - I guess there's no need to return `List` at all since element order is going to be randomized by `toMap`.

Reply:
I think so. Order of columns in `atomic` is a **very good** thing to have, but:

1. I don't think it is preserved in current solution anyway (due initial `JsonObject` transformation)
2. For current purposes it should be fine to have `Map`

Maybe we can define a separate method that takes data from `atomic` and orders them according to `Event`. This will be super-useful in RDB Shredder. If you have an idea how to do this *efficiently* - please do, if not - treat it as low priority.